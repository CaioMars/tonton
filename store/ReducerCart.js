import { createSlice } from "@reduxjs/toolkit";

export const cartSlice = createSlice({
  name: "cart",
  initialState: {
    items: [],
    subtotal: 0,
    delivery: 0,
    discount: 0,
    total: 0,
  },
  reducers: {
    add: (state, action) => {
      const { payload } = action;
      state.subtotal += payload.price * payload.quantity;
      state.total = state.subtotal + state.delivery - state.discount;
      state.items.push(payload);
    },
    remove: (state, action) => {
      const item = state.items[action.payload];
      if (typeof item === "undefined") return;
      state.subtotal -= item.price * item.quantity;
      state.total = state.subtotal + state.delivery - state.discount;
      state.items = state.items.filter(
        (item, index) => index !== action.payload
      );
    },
  },
});

export const { add, remove } = cartSlice.actions;

export default cartSlice.reducer;
