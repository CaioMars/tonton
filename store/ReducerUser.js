import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  uid: "",
  username: "",
  email: "",
  firstName: "",
  lastName: "",
  lastLogin: "",
  refreshToken: "",
  orders: [],
};

export const userSlice = createSlice({
  name: "user",
  initialState,
  reducers: {
    signin: (state, action) => {
      state = action.payload;
    },
    signout: (state, action) => {
      state = initialState;
    },
  },
});

export const { signin, signout } = userSlice.actions;

export default userSlice.reducer;
