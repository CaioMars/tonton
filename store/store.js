import { configureStore } from "@reduxjs/toolkit";
import cartReducer from "./ReducerCart";
import userReducer from "./ReducerUser";

export default configureStore({
  reducer: {
    cart: cartReducer,
    user: userReducer,
  },
});
