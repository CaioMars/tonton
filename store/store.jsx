import { configureStore } from "@reduxjs/toolkit";
import cartReducer from "./ReducerCart";

export default configureStore({
  reducer: {
    cart: cartReducer,
  },
});
