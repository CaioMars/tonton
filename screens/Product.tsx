import { Text, View } from "../components/Themed";
import { StyleSheet, Image, ScrollView } from "react-native";
import { Button } from "react-native-elements";
import { useDispatch } from "react-redux";
import { add } from "../store/ReducerCart";
import StarRating from "../components/StarRating";
import { ItemType } from "../types";

export default function Product({ navigation, route }: any) {
  const item: ItemType = route.params.item;
  const dispatch = useDispatch();

  const MAX_TITLE_LENGTH = 25;
  navigation.setOptions({
    title:
      item.title.length > MAX_TITLE_LENGTH
        ? `${item.title.substr(0, MAX_TITLE_LENGTH).trim()}...`
        : item.title,
  });

  return (
    <>
      <View style={styles.container}>
        <ScrollView style={styles.containerScrollView}>
          <Image
            source={{ uri: item.image }}
            style={styles.image}
            resizeMode="contain"
          />
          <View style={styles.containerContent}>
            <Text style={styles.title}>{item.title}</Text>
            <Text style={styles.description}>{item.description}</Text>
            <StarRating
              rating={item.rating?.rate || 0}
              count={item.rating?.count || 0}
            />
            <Text style={styles.price}>$ {item.price}</Text>
          </View>
        </ScrollView>
      </View>
      <View style={styles.containerBottom}>
        <Button
          title="Buy"
          onPress={() => {
            dispatch(add(item));
            navigation.replace("Cart");
          }}
        />
      </View>
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  containerScrollView: {},
  containerContent: {
    padding: 10,
  },
  containerBottom: {
    padding: 10,
  },
  image: {
    width: "100%",
    height: 300,
    padding: 40,
  },
  title: {
    fontSize: 20,
    fontWeight: "bold",
    marginBottom: 20,
  },
  description: {
    marginBottom: 20,
  },
  price: {
    fontSize: 24,
    fontWeight: "bold",
    marginBottom: 20,
  },
  separator: {
    marginVertical: 30,
    height: 1,
    width: "80%",
  },
});
