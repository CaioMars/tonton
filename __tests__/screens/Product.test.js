jest.useFakeTimers();
import * as React from "react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import store from "../../store";
import Product from "../../screens/Product";

it(`renders correctly`, async () => {
  const tree = await renderer
    .create(
      <Provider store={store}>
        <Product
          route={{
            params: {
              item: {
                id: 1,
                title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                price: 109.95,
                description:
                  "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                category: "men's clothing",
                image:
                  "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                rating: { rate: 3.9, count: 120 },
              },
            },
          }}
        />
      </Provider>
    )
    .toJSON();

  await expect(tree).toMatchSnapshot();
});
