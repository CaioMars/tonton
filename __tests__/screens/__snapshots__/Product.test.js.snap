// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> renders correctly 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <RCTScrollView
      style={Object {}}
    >
      <View>
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
            }
          }
          style={
            Object {
              "height": 300,
              "padding": 40,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              Object {
                "padding": 10,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                },
                Object {
                  "fontSize": 20,
                  "fontWeight": "bold",
                  "marginBottom": 20,
                },
              ]
            }
          >
            Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                },
                Object {
                  "marginBottom": 20,
                },
              ]
            }
          >
            Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "marginBottom": 10,
              }
            }
          >
            <Text />
            <Text />
            <Text />
            <Text />
            <Text />
            <Text
              style={
                Object {
                  "fontWeight": "bold",
                  "marginHorizontal": 5,
                }
              }
            >
              3.9
            </Text>
            <Text
              style={Object {}}
            >
              (
              120
              )
            </Text>
          </View>
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                },
                Object {
                  "fontSize": 24,
                  "fontWeight": "bold",
                  "marginBottom": 20,
                },
              ]
            }
          >
            $ 
            109.95
          </Text>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {},
              ]
            }
          >
            <View
              style={
                Object {
                  "alignSelf": "baseline",
                  "borderColor": "#999",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#ddd",
                        "justifyContent": "center",
                        "paddingHorizontal": 18,
                        "paddingVertical": 8,
                      },
                      Object {
                        "borderBottomLeftRadius": 5,
                        "borderTopLeftRadius": 5,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#242424",
                        "fontSize": 18,
                      }
                    }
                  >
                    -
                  </Text>
                </View>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "paddingHorizontal": 20,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#242424",
                      "fontSize": 14,
                    }
                  }
                >
                  1
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#ddd",
                        "justifyContent": "center",
                        "paddingHorizontal": 18,
                        "paddingVertical": 8,
                      },
                      Object {
                        "borderBottomRightRadius": 5,
                        "borderTopRightRadius": 5,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#242424",
                        "fontSize": 18,
                      }
                    }
                  >
                    +
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>,
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "padding": 10,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderRadius": 3,
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 3,
          },
          undefined,
          false,
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Buy
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
